### ISSUE TITLE
mkcert_caroot does not support '$HOME' or '~'

### ISSUE THREAD

**Issue by scrummer** [drud/ddev#3145]
**Describe the bug**
In our company, we want to provide a global config template which is preconfigured with some stuff like env-vars and omitted containers.
Therefore we want to get rid of the absolute path in `mkcert_caroot`.

I've tried to set the value to `$HOME/Library/Application Support/mkcert` and `~/Library/Application Support/mkcert`. However, both versions result in an error when trying to start a project.

We all work in a similar environment on MacOS.

**To Reproduce**
1. Run `ddev poweroff`
2. Set `mkcert_caroot` to either `$HOME/Library/Application Support/mkcert` or `~/Library/Application Support/mkcert` in your `~/.ddev/global_config.yaml`.
3. In a project, run `ddev start`

**Expected behavior**
The projects should start as the tilde (~) is a valid character for resolving the home directory on all *nix systems (as well as the `$HOME` variable). It should not produce any problems.

**Screenshots**
![image](https://user-images.githubusercontent.com/15173170/128472549-2689adf2-c4bb-4e57-8200-66294ddc2587.png)

**Version and configuration information (please complete the following information):**

- Host computer OS and Version: macOS Big Sur 11.5
- Docker Desktop version if on macOS or Windows: Version 3.5.2 (3.5.2.18)
- ddev version information:
  - architecture      	arm64
  - db                	drud/ddev-dbserver-mariadb-10.3:v1.17.3
  -  dba               	phpmyadmin:5
  - ddev-ssh-agent    	drud/ddev-ssh-agent:v1.17.0
  - docker            	20.10.7
  - docker-compose    	1.29.2
  - os                	darwin
  - router            	drud/ddev-router:v1.17.6
  - web               	drud/ddev-webserver:v1.17.6
- ~~config.yaml contents for the misbehaving project (the part before the comments)~~
- ~~Do you have any custom configuration (nginx, php, mysql) in the .ddev folder? If so, have you tried without them?~~

**Additional context**
Maybe an important thing to know: This issue was caused on a MacBook Pro **M1**, but was not tested on any Intel Mac.

**Comment by plexmaniax** [comment]
+1

**Comment by rfay** [comment]
That's a good idea, and PR is welcome.

Of course $HOME and ~ are shell features, only supported directly where the shell is involved. But it seems like there ought to be a good way to do this. 

mkcert_caroot is actually the output of `mkcert -CAROOT`, which of course uses the absolute path.

ddev uses this to see if mkcert has been installed, though, so I don't see any reason it couldn't be interpreted.

It wouldn't work at all though for cross-platform teams, and the CAROOT is in a different place on macOS than on Linux or Windows.

It's possible that ddev could use a different technique to know whether `mkcert -install` has been run.

**Comment by rfay** [comment]
There are other problems with checking in ~/.ddev/global_config.yaml. Even though I intended for it to be source-controllable, it got out of hand at some point, and since current projects are stored in there, it's not really easy to control.

How were you going to work around that? Just check in a global_config.yaml with no projects in it?

This is interesting to me because of course I had thought some people would want to check in the global config, although you're the first I've heard of in these years it's existed.

**Comment by scrummer** [comment]
Yeah, you're right - currently we work with a all-custom setup. Custom development-images, custom local setup with jwilder/nginx-proxy, etc...
Not only that this is all a mess to maintain, the setup from ddev is was we love because it just works. The whole project-setup is working fine so far but as we have many projects and not only one big project we're working on, we want to focus on a more or less generic solution (like a template) we want to be able to use for *most* projects. Of course (not only with the flexibility of ddev) we're open to change the setup on behalf of project-specific requirements.

Some issues we have faced so far:
* We would \*love\* to have hooks not only on project-level, also to be configured in the `global_config.yaml` (so global hooks which would work in every project).
* As you said, there are some user-specific configurations in the `global_config.yaml` like `project_info`, `mkcert_caroot` or `last_started_version`.

So our goal (to clearify) is:
* We want to provide a repository with all the basic `~/.ddev` configuration. So developers should be able to just checkout that repo with `git clone xy ~/.ddev` and can start working on projects.
* We have created a small set of default tools we use as `webimage` to provide them as we use 3 or 4 different main technologies in most of our projects.
* When something changes in our default stack (like there's a new package requirement in one of our used technologies) we want to be able to update that stuff in a centralized place and the devs just run `git pull` and they're up to date.

Some open points we don't really came to a conclusion yet are:
* We currently work with a self-written sync tool, to sync assets and databases between dev, staging and production. This tool is currently being installed in our custom docker images. There we probably try to migrate to global ddev commands.
* Per project, we currently have a `project.yml` where we define static property information like php-version, docroot, deploy servers/paths, etc. => Here we definitely want to migrate to gitlab variables only but have not decided yet, how we integrate that stuff with ddev as we must be able do work with those values when syncing assets/databases

So - I would love to work together you to make some changes to ddev, that we and of course the community will be able to also make use of these advantages. Currently ddev is really more than perfect to be used with single-projects (also big projects) but it's not quite suitable for being used in companies which have many customers and installations - companies that are dependent of a centralized maintainable solution.

I hope this (tbh) wall of text gives you a bit of a better view on what we like to achieve.

BR 
~ Scrummer

**Comment by scrummer** [comment]
And regarding the point "working together with you"; I'd love to propose some changes to ddev in a PR but unfortunately I'm not really into Go - and that, I think, will end in a mess when I start contributing some things 😄 
But if you're open to that - why not...

**Comment by rfay** [comment]
You can propose specific feature requests in a PR and see if they get traction.

A lot of teams are successfully using the very mature approach already there.

Some teams use composer to install specific ddev features, so they can install the same thing on each project. They may use docker-compose.*.yaml or config.*.yaml via composer (or another technique)

You could, of course, maintain your own webimage, but it's easier to add a `webimage_extra_packages` to the config.yaml to add them, or add them via a custom .ddev/web-build/Dockerfile.

Lots of techniques like the composer approach ought to work for you. As I said, specific feature requests are welcome. But reworking ddev to fit your team's existing workflow isn't really a path that works for a project like this.

**Comment by rfay** [comment]
Some companies have been sponsoring ddev (which has no current financial support) by having me consult with them to help with their processes with ddev, so that's another option. I think we could also find others who could do that.

**Comment by scrummer** [comment]
> As I said, specific feature requests are welcome. But reworking ddev to fit your team's existing workflow isn't really a path that works for a project like this.

Sure - that was not really our idea.

Thank you for those approaches! We will sit together to discuss a bit what and how we do and hopefully can bring something back into ddev 👍🏻

**Comment by rfay** [comment]
What I'd actually like to do with this is to not store the CAROOT in there at all, and to move the project list out of that file. I don't think the storing of CAROOT there is actually useful, it may only be used to mark that the CA has been created.

**Comment by rfay** [comment]
Really, we should add the capability to have a `global_config.*.yaml` and then you could put your team's override values in there and not worry about the rest of the global_config.yaml at all. Just like in the project config. Not sure why it's not already like that. But a global_config.team.yaml could contain your teams standard global config and override anything in global_config.yaml.
